### ISSUE TITLE
Performance issue with complex code flow graph

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#3081]
This is a tracking bug. The original was reported by @crankedguy in [this thread](https://github.com/microsoft/pylance-release/issues/2382).

**Comment by erictraut** [comment]
I've narrowed the issue down to the `tag_release` function. This function is extremely complex — one of the most complex code flow graphs that I've ever seen in a Python function. It has 54 local variables, none of which have annotated types. That means the type analyzer needs to infer all of their types. The main body of the function is nested 4 or 5 levels deep inside of loops and conditional blocks. The function also contains try/except blocks with continue statements within the except handlers. Many of the variable types depend on other variable types, so the type analyzer needs to iterate many types to resolve their types. The performance issue that you're seeing is because of the combinatoric explosion of analysis paths needed to infer all of these variable types within this complex code flow graph.

Pyright (the static type checker that pylance is built upon) already contains logic to detect extremely complex code flow paths. If a function is sufficiently complex, pyright does not attempt to infer the return type of that function. In this case, the performance issue is triggered not from a return type inference request but because of a request to evaluate the types of local variables. (These types are required to determine the semantic token types in pylance.)

@crankedguy, I'll need to think more about how to handle this situation. II will probably need to teach the type analyzer to "give up" when it sees such a complex function. That could result in incorrect semantic token colors for some symbols, but that's probably preferable to the performance issue.

You can work around the issue by refactoring the `tag_release` function and some of the other complex functions in the same file into more manageable subroutines.

**Comment by erictraut** [comment]
Just for my future reference, the repro code can be found in this private repo: https://github.com/crankedguy/python_issue.

**Comment by crankedguy** [comment]
@erictraut Thanks for your quick analysis and response.
I am in the middle of refactoring this function, thus it might be the "performance gain" of over 2 minutes to like 20-30 seconds from yesterday to today.
The thing is I was nearly finished refactoring this one :) but I will give it a bit more time

**Comment by crankedguy** [comment]
do you want to keep access to this repo? as you posted you reference to it? because I have no need for it

**Comment by erictraut** [comment]
Yes, I'd appreciate it if you would retain the repo until this bug is fixed.

**Comment by crankedguy** [comment]
ok will do, thanks

**Comment by crankedguy** [comment]
@erictraut Just to give a heads up, this was really nasty in parts in terms of reuising names for different types etc. It was late, after cleaning the mess up pylance finishes in 2-3 secs now.

**Comment by erictraut** [comment]
That's good to hear. I'd still like to see if there are ways to mitigate the performance impact because other pyright users will inevitably hit a similar issue in the future. Thanks for your help in providing the repro.

**Comment by crankedguy** [comment]
Yea sure no problem, just tell me here when you are finished, so I can remove the repo afterwards.

**Comment by mhmdmodan** [comment]
If it helps, I'm running into the same issue and I've been able to narrow down to a specific function.

Running MacOS 11.6, VS Code version 1.64.2, Pylance version 2022.2.3. Interpreter is base conda for Python 3.9.5 with just sklearn v1.0.2 pip installed

I get Pylance using up 100-200% CPU in the following code:

```python
from sklearn.decomposition import PCA

predicted = [[4, 3, 5], [6, 3, 2]]
tsne = PCA(n_components=2)
transformed = tsne.fit_transform(predicted)
```

Process: open vs code, no project or anything. Create a new python file, type in the first 4 lines, everything is fine. Type out the last line, with the fit_transform method, and CPU jumps up. Don't even have to save the file

**Comment by erictraut** [comment]
@mhmdmodan, the issue you're seeing is not related to the other issue in this bug report even though the symptoms are superficially similar. I've found and fixed the problem that you're seeing.

**Comment by erictraut** [comment]
@crankedguy, I've introduced a performance optimization that allows the analysis of your original code file to be analyzed in 6 seconds rather than 18 seconds. I realize that 6 seconds is still a long time, but it's a pretty extreme code flow graph. If you have taken steps to refactor it into more manageable subroutines, the analysis time should drop significantly.

**Comment by crankedguy** [comment]
@erictraut Thanks for the follow-up. This is great news. As I am under your optimized 6 seconds already after refactoring this should be a great experience now :)

**Comment by erictraut** [comment]
This is addressed in 1.1.226, which I just published. It will also be included in the next release of pylance.
